code/datums/chatmessage.dm

```diff
@@ -555,6 +579,45 @@
 	var/duration = BALLOON_TEXT_TOTAL_LIFETIME(duration_mult)
 	fadertimer = addtimer(CALLBACK(src, PROC_REF(end_of_life)), duration, TIMER_STOPPABLE|TIMER_DELETE_ME, SSrunechat)
 
+/**
+ * Approximates the chatmesseges width based on cached widths of each char.
+ * If the character is not found in this cache we assume the worst and add the highest possible value.
+ *
+ * Arguments:
+ * * string - string to measure width
+ * * font size - font size that the displayed string will be in, used to calculate font size multiplier
+ * * is_bold - passed if the font is bold, the approximation takes into account additional width of the font
+ * * has_icon - text has an icon, which adds extra 8 pixels
+ */
+/datum/chatmessage/proc/approx_str_width(string, font_size = DEFAULT_FONT_SIZE, is_bold = FALSE, has_icon = FALSE)
+	var/value = 0
+	var/index = NORMAL_FONT_INDEX
+	if(font_size == WHISPER_FONT_SIZE)
+		index = SMALL_FONT_INDEX
+	else if(font_size == BIG_FONT_SIZE)
+		index = BIG_FONT_INDEX
+
+	var/i = 1
+	while(i <= length(string))
+		var/list/letters = SSrunechat.letters[string[i]]
+		//List wasnt initialized or was tampered with
+		if(letters == null)
+			//Replace with question mark
+			var/char_len = length(string[i])
+			string = splicetext(string, i, i += char_len, "?")
+			value += SSrunechat.letters["?"][index]
+			i -= char_len
+		else
+			value += letters[index]
+
+		i += length(string[i])
+
+	if(is_bold)
+		value += length(string)
+
+	if(has_icon)
+		value += CHAT_MESSAGE_ICON_SIZE
+	return list(value, string)
```

Comment:
Allocating lists in the return value is a bad idea, since its on a datum you don't need a pointer because you can just modify the message as a class scoped variable.

Reply:
I will change it to pointers once we hit 515, for now I'll do var on the message itself 👍